pkg/agent/nodeportlocal/k8s/controller.go

```diff
@@ -19,43 +19,97 @@ package k8s
 import (
 	"encoding/json"
 	"fmt"
+	"reflect"
 	"sync"
 	"time"
 
 	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/portcache"
 
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/util/wait"
 	clientset "k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/util/workqueue"
 	"k8s.io/klog"
 )
 
 const (
-	minRetryDelay = 2 * time.Second
-	maxRetryDelay = 120 * time.Second
-	numWorkers    = 4
+	controllerName = "AntreaAgentNPLController"
+	minRetryDelay  = 2 * time.Second
+	maxRetryDelay  = 120 * time.Second
+	numWorkers     = 4
 )
 
-type Controller struct {
-	portTable  *portcache.PortTable
-	kubeClient clientset.Interface
-	queue      workqueue.RateLimitingInterface
-	Ctrl       cache.Controller
-	CacheStore cache.Store
-	PodToIP    map[string]string
-	podIPLock  sync.RWMutex
+type NPLController struct {
+	portTable   *portcache.PortTable
+	kubeClient  clientset.Interface
+	queue       workqueue.RateLimitingInterface
+	podInformer cache.SharedIndexInformer
+	podLister   corelisters.PodLister
+	svcInformer cache.SharedIndexInformer
+	podToIP     map[string]string
+	podIPLock   sync.RWMutex
 }
 
-func NewNPLController(kubeClient clientset.Interface, pt *portcache.PortTable) *Controller {
-	nplCtrl := Controller{
-		kubeClient: kubeClient,
-		portTable:  pt,
-		PodToIP:    make(map[string]string),
+func NewNPLController(kubeClient clientset.Interface,
+	podInformer cache.SharedIndexInformer,
+	svcInformer cache.SharedIndexInformer,
+	resyncPeriod time.Duration,
+	pt *portcache.PortTable) *NPLController {
+	c := NPLController{
+		kubeClient:  kubeClient,
+		portTable:   pt,
+		podInformer: podInformer,
+		podLister:   corelisters.NewPodLister(podInformer.GetIndexer()),
+		svcInformer: svcInformer,
+		podToIP:     make(map[string]string),
 	}
-	nplCtrl.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "nodeportlocal")
-	return &nplCtrl
+
+	podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.enqueuePod,
+			DeleteFunc: c.enqueuePod,
+			UpdateFunc: func(old, cur interface{}) {
+				oldPod := old.(*corev1.Pod)
+				curPod := cur.(*corev1.Pod)
+				// Pod fields to watch for: ContainerPort, HostIP, PodIP,
+				// Labels, Annotations (NPLAnnotationKey).
+				if oldPod.ResourceVersion != curPod.ResourceVersion {
```

Comment:
@tnqn do you know if this check is still useful when setting resyncPeriod to 0?

Reply:
Might just be a unit test issue, but after removing this I was getting an update storm in the unit tests.

Reply:
It shouldn't receive any dummy update events if resync is disabled. Could you point which unit test receive an update storm? I think there might be something wrong that we should address to avoid unnecessary calls to `UpdateFunc` in case it really happens when actually running.

Reply:
yes, looks like the Pod updates (for removing annotations) was being called, while updating the NPL annotation in Service updates. Although I could reproduce this only in the UTs, tried the same set of steps on my kubernetes cluster and everything seemed to be working fine. 
I've fixed it by avoiding irrelevant Pod updates from the NPLController.